drivers/wiegand/wiegand.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (C) 2019 Hackerspace San Salvador
+ *
+ * This driver is a port for RIOT-OS based on the Wiegand protocol for
+ * Arduino implementation from monkeyboards.org.
+ *
+ * Original implementation available at:
+ *   https://github.com/monkeyboard/Wiegand-Protocol-Library-for-Arduino
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ */
+
+/**
+ * @ingroup     drivers_wiegand
+ * @{
+ *
+ * @file
+ * @brief       Driver for the Wiegand Protocol.
+ *
+ * @author      Mario Gomez <mario.gomez@teubi.co>
+ *
+ * @}
+ */
+
+#include "xtimer.h"
+#include "wiegand_params.h"
+
+/**
+ * @brief Handler for D0 signal in Wiegand protocol
+ *
+ */
+void wg_read_d0(void *arg)
+{
+  wiegand_t *dev = (wiegand_t *)arg;
+  // Increament bit count for Interrupt connected to D0
+  dev->wg_bit_count++;
+  // If bit count more than 31, process high bits
+  if (dev->wg_bit_count>31)
+  {
+    // shift value to high bits
+    dev->wg_card_temp_high |= ((0x80000000 & dev->wg_card_temp)>>31);
+    dev->wg_card_temp_high <<= 1;
+    dev->wg_card_temp <<=1;
+  }
+  else
+  {
+    // D0 represent binary 0, so just left shift card data
+    dev->wg_card_temp <<= 1;
+  }
+  // Keep track of last wiegand bit received
+  dev->wg_last_wiegand = xtimer_now_usec64();
+}
+
+/**
+ * @brief Handler for D1 signal in Wiegand protocol
+ *
+ */
+void wg_read_d1(void *arg)
+{
+  wiegand_t *dev = (wiegand_t *)arg;
+
+  // Increment bit count for Interrupt connected to D1
+  dev->wg_bit_count ++;
+
+  // If bit count more than 31, process high bits
+  if (dev->wg_bit_count>31)
+  {
+    // shift value to high bits
+    dev->wg_card_temp_high |= ((0x80000000 & dev->wg_card_temp)>>31);
+    dev->wg_card_temp_high <<= 1;
+    dev->wg_card_temp |= 1;
+    dev->wg_card_temp <<=1;
+  }
+  else
+  {
+    // D1 represent binary 1, so OR card data with 1 then
+    dev->wg_card_temp |= 1;
+    // left shift card data
+    dev->wg_card_temp <<= 1;
+  }
+  // Keep track of last wiegand bit received
+  dev->wg_last_wiegand = xtimer_now_usec64();
```

Comment:
since `dev->wg_last_wiegand` is 32 bit anyway, `xtimer_now_usec()` is enough.

Reply:
Yes. We can use only one 64 bit integer. I suppose that the original author of the algorithm wanted to make the code a little bit more readable when decoding the Mifare 34 format.

Let me do this little change.